File: part1/week2/week2.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

const integerNum = 100000

func main() {
	xs := readFileAndMakeArray("IntegerArray.txt")
	_, numInv := sortAndCountInv(xs)
	fmt.Println(numInv)
}

func sortAndCountInv(xs []int) (sortedArr []int, numInv int) {
	n := len(xs)

	if n == 0 || n == 1 {
		return xs, 0
	}

	leftxs, leftInv := sortAndCountInv(xs[:n/2])
	rightxs, rightInv := sortAndCountInv(xs[n/2:])
	sortedArr, splitInv := mergeAndCountSplitInv(leftxs, rightxs)
	return sortedArr, leftInv + rightInv + splitInv

}

func mergeAndCountSplitInv(left []int, right []int) (mergeArr []int, splitInv int) {
	n := len(left) + len(right)
	mergeArr = make([]int, 0, n)
	splitInv = 0
	leftIndex := 0
	rightIndex := 0
	for i := 0; i < n; i++ {
		if rightIndex == len(right) || (leftIndex < len(left) && left[leftIndex] < right[rightIndex]) {
			mergeArr = append(mergeArr, left[leftIndex])
			leftIndex++
		} else {
			mergeArr = append(mergeArr, right[rightIndex])
			splitInv += len(left) - leftIndex
			rightIndex++
		}
	}
	return mergeArr, splitInv
}

func readFileAndMakeArray(name string) (xs []int) {
	xs = make([]int, 0, integerNum)
	f, err := os.Open(name)
	if err != nil {
		log.Fatalln(err)
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line, err := strconv.Atoi(scanner.Text())
		if err != nil {
			log.Fatalln(err)
		}
		xs = append(xs, line)
	}
	return xs
}

File: part1/week2/week2.py
FILENAME = 'IntegerArray.txt'

def readFileAndCreateArray():
    file = open(FILENAME, 'r')
    lines = file.read().splitlines()
    file.close()
    numbers = []
    for line in lines:
        numbers.append(int(line))
    return numbers

def mergeAndContSplitInv(leftArr, rightArr):
    mergedArr = []
    splitInv = 0
    n = len(leftArr) + len(rightArr)
    leftIndex  = 0
    rightIndex = 0
    for _ in range(n):
        if rightIndex == len(rightArr) or (leftIndex < len(leftArr) and leftArr[leftIndex] < rightArr[rightIndex]):
            mergedArr.append(leftArr[leftIndex])
            leftIndex += 1
        else:
            mergedArr.append(rightArr[rightIndex])
            splitInv   += (len(leftArr) - leftIndex) 
            rightIndex += 1
    return mergedArr, splitInv
    

def sortAndCountInv(arr):
    n = len(arr)
    # base case
    if n == 0 or n == 1:
        return arr, 0
    # n >= 2
    leftArr, leftInv   = sortAndCountInv(arr[:n//2])
    rightArr, rightInv = sortAndCountInv(arr[n//2:])
    mergedArr, splitInv = mergeAndContSplitInv(leftArr, rightArr)
    return mergedArr, leftInv + rightInv + splitInv



if __name__ == '__main__':
    numbers = readFileAndCreateArray()
    numbers, ans = sortAndCountInv(numbers)
    print(ans)
